### PR TITLE
Add turbo turtle

### PR DESCRIPTION
**Pull request by kbrindley**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by kbrindley** [comment]
@tprobertslanl are you willing to be listed as a maintainer on the conda-forge package?

**Comment by tprobertslanl** [comment]
Yes, I'm willing to be a maintainer

**Comment by kbrindley** [comment]
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1025658&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1025658&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1240

```
error: package directory 'turbo_turtle/tutorials' does not exist
```

problem with the conda-build/setuptools versions and the pyproject.toml file. I think this will require a version bump to fix 😢

**Comment by kbrindley** [comment]
@conda-forge/help-python

I think this is basically ready to go. I'm not sure why the Windows job is running long.

I've made all the patches mentioned in the TODO notes on the upstream repository, but the internal release process will take a month or two before the updated package hits GitHub. I'd like to release the package/recipe as-is in the meantime.

Cheers,

Kyle

File: recipes/turbo_turtle/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+build:
+  noarch: python
+  number: 0
+  script_env:
+    - SETUPTOOLS_SCM_PRETEND_VERSION={{ version }}
+  entry_points:
+    - turbo-turtle = turbo_turtle._main:main
+  ignore_run_exports:
+    - python_abi
```
Comment (xhochy) [inline recipes/turbo_turtle/meta.yaml:20]: Why are you doing this?

File: recipes/turbo_turtle/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+requirements:
+  build:
```
Comment (xhochy) [inline recipes/turbo_turtle/meta.yaml:23]: These should all be in host

File: recipes/turbo_turtle/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    - lazy_loader >=0.4
+    - scons >=4
+    - setuptools_scm[toml] >=6.2
```
Comment (xhochy) [inline recipes/turbo_turtle/meta.yaml:26]: Things in square brackets are basically ignored. Thia doesn't do what you expect.

File: recipes/turbo_turtle/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    - python >=3.8
+    - lazy_loader >=0.4
+    - scons >=4
+    - waves >=0.7.2,<0.11.0
+
+test:
+  requires:
+    - pytest
+    - pytest-xdist
+  imports:
```
Comment (xhochy) [inline recipes/turbo_turtle/meta.yaml:49]: Please also run pip check in the tests.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: recipes/turbo_turtle/meta.yaml (hunk 1 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:20]: https://github.com/conda/conda-build/issues/3956#issuecomment-845711269

File: recipes/turbo_turtle/meta.yaml (hunk 2 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:23]: I'm using the build/host separation for PREFIX/BUILD_PREFIX separation and trying to follow the build/host separation in the conda-build documentation: https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#requirements

Am I interpreting the build/host separation incorrectly for a noarch package? Does conda-forge have documented preferences for build/host section use for pure Python packages?

File: recipes/turbo_turtle/meta.yaml (hunk 3 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:26]: This used to be the recommended requirement from the setuptools_scm package. Poking around a little deeper, it looks like this was always a PIP syntax and never a Conda syntax.

It looks like they've updated their documentation: https://setuptools-scm.readthedocs.io/en/stable/#with-setuptools. I can update the recipe and upstream package to reflect the current docs.

File: recipes/turbo_turtle/meta.yaml (hunk 4 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:49]: This isn't a PyPI package. It's being built directly from the GitHub release source. Will ``pip check`` have any affect?

https://conda-forge.org/docs/maintainer/adding_pkgs/#pip-check

**Comment by kbrindley** [comment]
> Hi! This is the friendly automated conda-forge-linting service.
> 
> I Failed to even lint the recipe, probably because of a conda-smithy bug 😢. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint --conda-forge .` from the recipe directory.

```
% pwd
/.../staged-recipes/recipes/turbo_turtle
% conda smithy recipe-lint --conda-forge .
. is in fine form
```

File: recipes/turbo_turtle/meta.yaml (hunk 4 above)
Comment (xhochy) [reply recipes/turbo_turtle/meta.yaml:49]: Yes, but it is a Python package and `pip check` only checks for Python package metadata consistency and doesn't even contact PyPI for that.

File: recipes/turbo_turtle/meta.yaml (hunk 2 above)
Comment (xhochy) [reply recipes/turbo_turtle/meta.yaml:23]: Generally, Python packages should be in a host as they are specific to the target. In the `noarch` case, `build` should be empty`

File: recipes/turbo_turtle/meta.yaml (hunk 4 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:49]: Got it. I've made the change and push soon. For my own understanding, ``pip check`` understands how to search and check packages installed in a conda environment, not just pip environments?

File: recipes/turbo_turtle/meta.yaml (hunk 1 above)
Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:20]: Unless behavior has changed in conda-build, the package will inherit an unnecessary Python ABI dependency when building against the conda-forge channel.

I'll try running without it to see if the runtime dependencies still look correct.

Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml]: looks like the behavior has changed. I've removed these lines.

File: recipes/turbo_turtle/meta.yaml (hunk 4 above)
Comment (xhochy) [reply recipes/turbo_turtle/meta.yaml:45]: It only understands pip environments, but conda environments with pip installed are also "pip environments" metadata-wise. This only checks the subset of dependency constraints that is exposed to pip. We use this to partially ensure that pip and conda constraints are inline.

Comment (kbrindley) [reply recipes/turbo_turtle/meta.yaml:45]: awesome. Thanks!

**Comment by kbrindley** [comment]
@xhochy I think I've addressed your comments. Do you prefer to resolve the review conversations yourself or should I?

**Comment by xhochy** [comment]
You still need to address the linter, then this is good to go.

**Comment by kbrindley** [comment]
@xhochy thanks. I missed the linter comment because the tests passed. Looks like I accidentally removed setuptools when merging the build/host sections.

**Review by xhochy — APPROVED** [review]

**Pull request merged**